relay-event-schema/src/protocol/span/convert.rs

```diff
@@ -0,0 +1,254 @@
+//! This module defines bidirectional field mappings between spans and transactions.
+use crate::protocol::{
+    ContextInner, Contexts, DefaultContext, Event, ProfileContext, Span, TraceContext,
+};
+
+use relay_protocol::Annotated;
+use std::collections::BTreeMap;
+
+/// Implements the conversion between transaction events and segment spans.
+///
+/// Invoking this macro implements both `From<&Event> for Span` and `From<&Span> for Event`.
+///
+/// Example:
+///
+/// ```skip
+/// map_fields!(
+///     top-level:
+///         span.received: event.received
+///     contexts:
+///         TraceContext:
+///             span.trace_id: context.trace_id
+///         ;
+///     ;
+///     fixed_for_span:
+///         // ...
+///     ;
+///     fixed_for_event:
+///         // ...
+///     ;
+/// );
```

Comment:
I would have preferred a flat mapping like this:

```rust
map_fields!(
  span.received <=> event.received,
  span.trace_id <=> event.contexts.trace.trace_id,
  span.is_segment <= Some(true)
);
```

but I could not make it work with a simple macro, and a procedural macro seemed overkill for this purpose.

Reply:
I think we can make that work if it is just about `<=>`, `=>`, `<=`, do you want me take a look?

Reply:
There's two problems:

1. You need to cast a context to a specific type to access its fields.
2. The compiler will complain about parsing ambiguity on the LHS identifier when there's no separators between the `<=>` and `<=`  blocks.

But I'm very open to ideas, there's probably possibilities I haven't discovered yet!

Reply:
Yeah contexts would need to stay separate (or with special syntax), but maybe it's possible to combine `top-level` `fixed-for-span` and `fixed-for-event`.